### PR TITLE
Import state sign convention change and test improvements

### PR DESCRIPTION
**Pull request by Chiffafox**
**Context:**

Initial state preparation in PennyLane now imports states based on the physicist sign convention (interleaved operators). This allows exact agreement when energies are evaluated for the same state using PennyLane's Hamiltonian matrix, leading to much better acceleration of quantum algorithms.

**Description of the Change:**

A sign convention is applied to initial states that originated in PySCF (CISD/CCSD) or Dice (SHCI) to convert them to physicist convention in PennyLane (interleaved spin-up / spin-down operators). Additionally, new tests are introduced comparing the energy of the initial state from the method which contributed the state, to that same state's energy as evaluated in PennyLane. This is an extra check ensuring that the state import process is correct. 

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Review by soranjh — CHANGES_REQUESTED** [review]

File: pennylane/qchem/convert.py [hunk 1]
```diff
@@ -1202,4 +1214,22 @@ def _shci_state(wavefunction, tol=1e-15):
...
 
+    # convert sign parity from chemist to physicist convention (interleaving spin operators)
+    dict_fcimatr = _sign_chem_to_phys(dict_fcimatr, norb)
+
     return dict_fcimatr
+
+
+def _sign_chem_to_phys(dict_fcimatr, norb):
+    # need to compute the permutation sign in passing from
+    # chemist to physicist convention of storing spin operators as interleaved
```
[the remark this answers is not in the document]
Comment (Chiffafox) [reply pennylane/qchem/convert.py]: Done.

File: pennylane/qchem/convert.py [hunk 2]
```diff
@@ -745,15 +758,23 @@ def _rcisd_state(cisd_solver, tol=1e-15):
 
     # alpha, beta -> alpha, beta excitations
     # generate all possible pairwise combinations of _single_ excitations of alpha and beta sectors
-    rowvals, colvals = np.array(list(product(c1a_configs, c1a_configs)), dtype=int).T
-    c2ab = (c2.transpose(0, 2, 1, 3).reshape(nocc * nvir, -1)).ravel()
+    rowvals, colvals = np.array(list(product(c1a_configs, c1a_configs)), dtype=int).T.numpy()
```
[the remark this answers is not in the document]
Comment (Chiffafox) [reply pennylane/qchem/convert.py]: That's great! Removed those.

File: pennylane/qchem/convert.py [hunk 3]
```diff
@@ -1202,4 +1214,22 @@ def _shci_state(wavefunction, tol=1e-15):
     # filter based on tolerance cutoff
     dict_fcimatr = {key: value for key, value in dict_fcimatr.items() if abs(value) > tol}
 
+    # convert sign parity from chemist to physicist convention (interleaving spin operators)
+    dict_fcimatr = _sign_chem_to_phys(dict_fcimatr, norb)
+
     return dict_fcimatr
+
+
```
[the remark this answers is not in the document]
Comment (Chiffafox) [reply pennylane/qchem/convert.py:1180]: Done.

File: pennylane/qchem/convert.py [hunk 4]
```diff
@@ -746,14 +657,124 @@ def _rcisd_state(cisd_solver, tol=1e-15):
     # alpha, beta -> alpha, beta excitations
     # generate all possible pairwise combinations of _single_ excitations of alpha and beta sectors
     rowvals, colvals = np.array(list(product(c1a_configs, c1a_configs)), dtype=int).T
-    c2ab = (c2.transpose(0, 2, 1, 3).reshape(nocc * nvir, -1)).ravel()
+    c2ab = c2.transpose(0, 2, 1, 3).reshape(nocc * nvir, -1)
     dict_fcimatr.update(
-        dict(zip(list(zip(rowvals, colvals)), c2ab * np.kron(c1a_signs, c1a_signs)))
+        dict(
+            zip(
+                list(zip(rowvals, colvals)),
```
Comment (obliviateandsurrender) [inline pennylane/qchem/convert.py:664]: At first glance, it feels like this is unnecessary. You might be able to get the same thing by directly using `list(product(c1a_configs, c1a_configs)` (or something similar), which will help you skip defining `rowvals` and `colvals`.

File: pennylane/qchem/convert.py [hunk 5]
```diff
@@ -746,14 +657,124 @@ def _rcisd_state(cisd_solver, tol=1e-15):
...
+    nelec = mol.nelectron
+    nelec_a = int((nelec + mol.spin) / 2)
+    nelec_b = int((nelec - mol.spin) / 2)
+
+    nvir_a, nvir_b = norb - nelec_a, norb - nelec_b
+
+    size_a, size_b = nelec_a * nvir_a, nelec_b * nvir_b
+    size_aa = int(nelec_a * (nelec_a - 1) / 2) * int(nvir_a * (nvir_a - 1) / 2)
+    size_bb = int(nelec_b * (nelec_b - 1) / 2) * int(nvir_b * (nvir_b - 1) / 2)
+    size_ab = nelec_a * nelec_b * nvir_a * nvir_b
```
Comment (obliviateandsurrender) [inline pennylane/qchem/convert.py:727]:
```suggestion
    size_ab = size_a * size_b
```

File: pennylane/qchem/convert.py [hunk 6]
```diff
@@ -746,14 +657,124 @@ def _rcisd_state(cisd_solver, tol=1e-15):
...
+    c2aa_configs, c2aa_signs = _excited_configurations(nelec_a, norb, 2)
+    dict_fcimatr.update(dict(zip(list(zip(c2aa_configs, [ref_b] * size_aa)), c2aa * c2aa_signs)))
+
+    # alpha, beta -> alpha, beta excitations
+    rowvals, colvals = np.array(list(product(c1a_configs, c1b_configs)), dtype=int).T
+    dict_fcimatr.update(
+        dict(
+            zip(
+                list(zip(rowvals, colvals)),
+                np.einsum("i,j,ij->ij", c1a_signs, c1b_signs, c2ab).ravel(),
```
Comment (obliviateandsurrender) [inline pennylane/qchem/convert.py:763]: Same comment as before regarding `np.einsum`'s efficiency and regarding having `rowvals` and `colvals`.

File: pennylane/qchem/convert.py [hunk 7]
```diff
@@ -877,7 +898,14 @@ def _rccsd_state(ccsd_solver, tol=1e-15):
     # alpha, beta -> alpha, beta excitations
     rowvals, colvals = np.array(list(product(t1a_configs, t1b_configs)), dtype=int).T
```
Comment (obliviateandsurrender) [inline pennylane/qchem/convert.py:899]: This looks like a third occurrence of similar code, i.e., building this dictionary form from method-specific configurations. We could clean up some code by making a separate helper function for doing this conversion all together which will help us condense repetitive code.

**Review by obliviateandsurrender — CHANGES_REQUESTED** [review]
Thanks @Chiffafox! Left some comments and suggestions. Don't forget to add a nice changelog entry for the change. 🦊

File: pennylane/qchem/convert.py (hunk 4 above)
Comment (Chiffafox) [reply pennylane/qchem/convert.py:664]: Good catch -- we transpose and separate it only to zip it back again. Fixed.

File: pennylane/qchem/convert.py [hunk 8]
```diff
@@ -746,14 +657,124 @@ def _rcisd_state(cisd_solver, tol=1e-15):
...
     # generate all possible pairwise combinations of _single_ excitations of alpha and beta sectors
     rowvals, colvals = np.array(list(product(c1a_configs, c1a_configs)), dtype=int).T
-    c2ab = (c2.transpose(0, 2, 1, 3).reshape(nocc * nvir, -1)).ravel()
+    c2ab = c2.transpose(0, 2, 1, 3).reshape(nocc * nvir, -1)
     dict_fcimatr.update(
-        dict(zip(list(zip(rowvals, colvals)), c2ab * np.kron(c1a_signs, c1a_signs)))
+        dict(
+            zip(
+                list(zip(rowvals, colvals)),
+                np.einsum("i,j,ij->ij", c1a_signs, c1a_signs, c2ab).ravel(),
```
[the remark this answers is not in the document]
Comment (Chiffafox) [reply pennylane/qchem/convert.py:665]: Good point -- added optimize=True. Looked around and even though we can do the same with `np.outer` plus `.ravel()`, seems that `einsum` is often faster than numpy's built-in functions. From experience, this has never been the bottleneck, anyway.

File: pennylane/qchem/convert.py [hunk 9]
```diff
@@ -746,14 +657,124 @@ def _rcisd_state(cisd_solver, tol=1e-15):
...
+    nvir_a, nvir_b = norb - nelec_a, norb - nelec_b
+
+    size_a, size_b = nelec_a * nvir_a, nelec_b * nvir_b
+    size_aa = int(nelec_a * (nelec_a - 1) / 2) * int(nvir_a * (nvir_a - 1) / 2)
+    size_bb = int(nelec_b * (nelec_b - 1) / 2) * int(nvir_b * (nvir_b - 1) / 2)
+    size_ab = nelec_a * nelec_b * nvir_a * nvir_b
+
+    sizes = [1, size_a, size_b, size_ab, size_aa, size_bb]
+    cumul = np.cumsum(sizes)
+    idxs = [0] + [slice(cumul[ii], cumul[ii + 1]) for ii in range(len(cumul) - 1)]
```
[the remark this answers is not in the document]
Comment (Chiffafox) [reply pennylane/qchem/convert.py:731]: It is indeed a numpy array. Thank you for the suggestion! Nice-looking and saves a few lines. :)

File: pennylane/qchem/convert.py (hunk 6 above)
Comment (Chiffafox) [reply pennylane/qchem/convert.py:763]: Done, see above.

File: pennylane/qchem/convert.py (hunk 7 above)
Comment (Chiffafox) [reply pennylane/qchem/convert.py:899]: We thought about this early on, but unfortunately there are enough small differences between the four occurrences you mention (RCISD/UCISD/RCCSD/UCCSD) that there is not a simple helper function to abstract this part away.

File: pennylane/qchem/convert.py [hunk 10]
```diff
@@ -1007,7 +1039,14 @@ def _uccsd_state(ccsd_solver, tol=1e-15):
     # alpha, beta -> alpha, beta excitations
     rowvals, colvals = np.array(list(product(t1a_configs, t1b_configs)), dtype=int).T
     dict_fcimatr.update(
-        dict(zip(list(zip(rowvals, colvals)), t2ab.ravel() * np.kron(t1a_signs, t1b_signs)))
+        dict(
+            zip(
+                list(zip(rowvals, colvals)),
+                np.einsum(
```
[the remark this answers is not in the document]
Comment (Chiffafox) [reply pennylane/qchem/convert.py:1045]: Done, see above.

File: pennylane/qchem/convert.py [hunk 11]
```diff
@@ -1202,4 +1176,84 @@ def _shci_state(wavefunction, tol=1e-15):
...
+    """
+
+    signed_dict = {}
+    for key, elem in dict_fcimatr.items():
+        lsta, lstb = bin(key[0])[2:][::-1], bin(key[1])[2:][::-1]
+        # highest energy state is on the right -- pad to the right
+        lsta = np.array([int(elem) for elem in lsta] + [0] * (norb - len(lsta)))
+        lstb = np.array([int(elem) for elem in lstb] + [0] * (norb - len(lstb)))
+        which_occ = np.where(lsta == 1)[0]
+        parity = (-1) ** np.sum([np.sum(lstb[: int(ind)]) for ind in which_occ])
```
[the remark this answers is not in the document]
Comment (Chiffafox) [reply pennylane/qchem/convert.py:1257]: I suppose it is, but having it there makes things more readable. It is counting the number of occurrences of the spin down operators to the left of spin-up operators, to see how many interchanges and thus negative signs are picked up. Then those signs are summed together to determine the global sign.

File: pennylane/qchem/convert.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Chiffafox) [reply pennylane/qchem/convert.py:1045]: Done, see above.

File: pennylane/qchem/convert.py (hunk 7 above)
Comment (obliviateandsurrender) [reply pennylane/qchem/convert.py]: I meant to condense these into a helper function - 
image

File: pennylane/qchem/convert.py (hunk 5 above)
Comment (Chiffafox) [reply pennylane/qchem/convert.py]: Thank you, done!

File: pennylane/qchem/convert.py (hunk 7 above)
Comment (Chiffafox) [reply pennylane/qchem/convert.py]: I agree that it is possible to replace with a helper, but that helper will need 5-6 arguments and in my mind breaks up the flow / readability of the function. I prefer to keep the `einsum` explicit.

**Review by obliviateandsurrender — APPROVED** [review]

File: doc/releases/changelog-dev.md [hunk 12]
```diff
@@ -76,12 +76,17 @@
 * `StatePrep` operations expanded onto more wires are now compatible with backprop.
   [(#5028)](https://github.com/PennyLaneAI/pennylane/pull/5028)
 
+* `import_state` now applies the chemist-to-physicist sign convention switch when
```
Comment (obliviateandsurrender) [inline doc/releases/changelog-dev.md:79]: might want to add a link for `chemist-to-physicist` sign convention.

**Review by soranjh — CHANGES_REQUESTED** [review]
Thanks @Chiffafox, could you please double-check the code examples for the following functions and verify/update the consistency of the outputs? I get different signs. 

1. `_rccsd_state`
2. `_uccsd_state`

File: doc/releases/changelog-dev.md (hunk 12 above)
Comment (Chiffafox) [reply doc/releases/changelog-dev.md]: Couldn't find a simple link to this convention, but expanded the explanation and also comments within the functions themselves to make it clear (also the docstring of _sign_chem_to_phys` has a more in-depth explanation).

**Pull request closed without merging**